ruby/ql/lib/codeql/ruby/typetracking/ApiGraphShared.qll

```diff
@@ -0,0 +1,329 @@
+private import TypeTrackerSpecific
+
+/**
+ * The signature to use when instantiating `ApiGraphShared`.
+ *
+ * The implementor should define a newtype with at least three branches as follows:
+ * ```ql
+ * newtype TApiNode =
+ *   MkForwardNode(LocalSourceNode node, TypeTracker t) { isReachable(node, t) } or
+ *   MkBackwardNode(LocalSourceNode node, TypeTracker t) { isReachable(node, t) } or
+ *   MkSinkNode(Node node) { ... } or
+ *   ...
+ * ```
+ *
+ * The three branches should be exposed through `getForwardNode`, `getBackwardNode`, and `getSinkNode`, respectively.
+ */
+signature module ApiGraphSharedSig {
+  /** A node in the API graph. */
+  class ApiNode {
+    /** Gets a string representation of this API node. */
+    string toString();
+
+    /** Gets the location associated with this API node, if any. */
+    Location getLocation();
+  }
+
+  /**
+   * Gets the forward node with the given type-tracking state.
+   *
+   * This node will have outgoing epsilon edges to its type-tracking successors.
+   */
+  ApiNode getForwardNode(TypeTrackingNode node, TypeTracker t);
+
+  /**
+   * Gets the backward node with the given type-tracking state.
+   *
+   * This node will have outgoing epsilon edges to its type-tracking predecessors.
+   */
+  ApiNode getBackwardNode(TypeTrackingNode node, TypeTracker t);
+
+  /**
+   * Gets the sink node corresponding to `node`.
+   *
+   * Since sinks are not generally `LocalSourceNode`s, such nodes are materialised separately in order for
+   * the API graph to include representatives for sinks. Note that there is no corresponding case for "source"
+   * nodes as these are represented as forward nodes with initial-state type-trackers.
+   *
+   * Sink nodes have outgoing epsilon edges to the backward nodes corresponding to their local sources.
+   */
+  ApiNode getSinkNode(Node node);
+
+  /**
+   * Holds if a language-specific epsilon edge `pred -> succ` should be generated.
+   */
+  predicate specificEpsilonEdge(ApiNode pred, ApiNode succ);
+}
+
+/**
+ * Parts of API graphs that can be shared between language implementations.
+ */
+module ApiGraphShared<ApiGraphSharedSig S> {
+  private import S
+
+  /** Gets a local source of `node`. */
+  bindingset[node]
+  pragma[inline_late]
+  TypeTrackingNode getALocalSourceStrict(Node node) { result = node.getALocalSource() }
+
+  cached
+  private module Cached {
+    /**
+     * Holds if there is an epsilon edge `pred -> succ`.
+     *
+     * That relation is reflexive, so `fastTC` produces the equivalent of a reflexive, transitive closure.
+     */
+    pragma[noopt]
+    cached
+    predicate epsilonEdge(ApiNode pred, ApiNode succ) {
+      // forward
+      exists(
+        StepSummary summary, TypeTrackingNode predNode, TypeTracker predState,
+        TypeTrackingNode succNode, TypeTracker succState
+      |
+        StepSummary::stepCall(predNode, succNode, summary)
+        or
+        StepSummary::stepNoCall(predNode, succNode, summary)
```

Comment:
```suggestion
        StepSummary::step(predNode, succNode, summary)
```

?

Reply:
We're in a `pragma[noopt]` context and so we can't call the `step` predicate which is `pragma[inline]`.

Predicates with `pragma[inline_late]` _can_ be called from `pragma[noopt]` and it's tempting to factor out such a helper predicate. However, such a helper predicate would need to have an empty binding set, since we're using it as the very first step of the pipeline, and @jbj keeps saying no when I ask for support for empty binding sets.

Reply:
I'll be happy to consider benchmarks that demonstrate the need. The current thinking in my team is that either it'll be cheap to call the predicate even when it's not inline, or the RA-level inlining will inline it even when it's not marked with any pragma.